middlewares/accesslog/jsonLogFormatter.go

```diff
@@ -0,0 +1,193 @@
+package accesslog
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"github.com/containous/traefik/log"
+	"github.com/containous/traefik/types"
+	"io"
+	"math"
+	"net/http"
+	"os"
+	"strconv"
+	"strings"
+	"time"
+)
+
+type tuple struct {
+	sourceKey, translatedKey string
+}
+
+type jsonLogFormatter struct {
+	timeFormat                       string
+	coreMapping, requestMapping      []tuple
+	originMapping, downstreamMapping []tuple
+}
+
+func convertFieldsToMappings(fields []string) []tuple {
+	var mapping []tuple
+	for _, s := range fields {
+		colon := strings.IndexByte(s, ':')
+
+		var t tuple
+
+		if colon > 0 {
+			t = tuple{
+				strings.TrimSpace(s[:colon]),
+				strings.TrimSpace(s[colon+1:]),
+			}
+		} else {
+			ws := strings.TrimSpace(s)
+			t = tuple{ws, ws}
+		}
+
+		mapping = append(mapping, t)
+	}
+	return mapping
+}
+
+func newJSONLogFormatter(settings *types.AccessLog) jsonLogFormatter {
+	jlf := jsonLogFormatter{timeFormat: settings.TimeFormat}
+
+	if len(settings.CoreFields) == 0 {
+		// default is to propagate all fields
+		for _, name := range defaultCoreKeys {
+			t := tuple{name, name}
+			jlf.coreMapping = append(jlf.coreMapping, t)
+		}
+	} else {
+		jlf.coreMapping = convertFieldsToMappings(settings.CoreFields)
+		validateCoreFields(jlf.coreMapping)
+	}
+
+	if len(settings.RequestHeaders) > 0 {
+		jlf.requestMapping = convertFieldsToMappings(settings.RequestHeaders)
+	}
+
+	if len(settings.OriginResponseHeaders) > 0 {
+		jlf.originMapping = convertFieldsToMappings(settings.OriginResponseHeaders)
+	}
+
+	if len(settings.DownstreamResponseHeaders) > 0 {
+		jlf.downstreamMapping = convertFieldsToMappings(settings.DownstreamResponseHeaders)
+	}
+
+	return jlf
+}
+
+func validateCoreFields(mappings []tuple) {
+	var invalidFields []string
+
+	for _, m := range mappings {
+		key := m.sourceKey
+		if _, exists := allCoreKeys[key]; !exists {
+			invalidFields = append(invalidFields, key)
+		}
+	}
+
+	if len(invalidFields) > 0 {
+		log.Errorf("Unsupported access log fields: %v", invalidFields)
+		exiter.Exit(1)
+	}
+}
+
+//-------------------------------------------------------------------------------------------------
+
+// Exiter provides a facade for os.Exit.
+type Exiter interface {
+	Exit(code int)
+}
+
+type stdExiter struct{}
+
+func (stdExiter) Exit(code int) {
+	os.Exit(code)
+}
+
+// exiter provides a seam for testing code containing os.Exit.
+var exiter Exiter = stdExiter{}
+
+//-------------------------------------------------------------------------------------------------
+
+func (l jsonLogFormatter) Write(w io.Writer, logDataTable *LogData) error {
+
+	buf := &bytes.Buffer{}
+	buf.WriteString("{")
```

Comment:
I see that we are producing JSON (braces here, objects below) manually. Can we try to populate our content in some map/struct and call `json.Marshal` on it instead?

Implementing `json.Marshaler` on some type might be helpful too.